File: countConnectedComponents.js
class Solution {
  /**
   * @param {number} n
   * @param {number[][]} edges
   * @returns {number}
   */
  countComponents(n, edges) {
    let res = 0;
    let adj = Array.from({ length: n }, () => []);
    let visited = Array.from({ length: n }, () => false);

    for (const [a, b] of edges) {
      adj[a].push(b);
      adj[b].push(a);
    }

    const dfs = (node) => {
      for (let edge of adj[node]) {
        if (!visited[edge]) {
          visited[edge] = true;
          dfs(edge);
        }
      }
    };

    for (let node = 0; node < n; node++) {
      if (!visited[node]) {
        visited[node] = true;
        dfs(node);
        res++;
      }
    }

    return res;
  }
}

File: count_connected_components.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        adj = [[] for node in range(n)]
        visit = [False] * n
        res = 0

        for a, b in edges:
            adj[a].append(b)
            adj[b].append(a)

        def dfs(node):
            for edge in adj[node]:
                if not visit[edge]:
                    visit[edge] = True
                    dfs(edge)

        for node in range(n):
            if not visit[node]:
                visit[node] = True
                dfs(node)
                res += 1

        return res
